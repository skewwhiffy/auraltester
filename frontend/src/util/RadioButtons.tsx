import React from 'react'
import { Form } from "react-bootstrap";

type OnChangeHandler = (value: string) => void

export interface RadioButtonDefinition {
  key?: string,
  value: string,
  label?: string
}

export interface Props {
  values: Array<RadioButtonDefinition>,
  name: string,
  defaultValue?: string,
  onChange: OnChangeHandler
}

interface State {
  value: string
}

class RadioButtons extends React.Component<Props, State> {
  constructor(props: Props) {
    if (props.values.length === 0) {
      throw Error('Need at least one radio button')
    }
    super(props);
  }

  private onChange = (ev: any) => {
    (this.props.onChange || (_ => { }))(ev.target.value)
  }

  private renderRadioButton = (radioButton: RadioButtonDefinition) => {
    const value = radioButton.value
    const key = radioButton.key || value
    const label = radioButton.label || value
    const defaultValue = this.props.defaultValue === undefined
      ? this.props.values[0].value
      : this.props.defaultValue
    return (
      <Form.Check
        id={`${this.props.name}-${key}`}
        key={key}
        name={this.props.name}
        value={value}
        label={label}
        type='radio'
        defaultChecked={value === defaultValue}
      />
    )
  }

  render() {
    return (
      <Form.Group onChange={this.onChange}>
        {this.props.values.map(this.renderRadioButton)}
      </Form.Group>
    )
  }
}

export default RadioButtons