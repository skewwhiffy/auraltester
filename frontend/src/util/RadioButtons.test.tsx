import renderer from 'react-test-renderer'
import RadioButtons from './RadioButtons'

describe('RadioButtons', () => {
  it('should match snapshot', () => {
    const values = Array.from(Array(5).keys())
    .map(it => ({
      value: `${it}-value`,
      key: `${it}-key`,
      label: `${it}-label`
    }))
    const onChange = (value: string) => {}

    const actual = renderer.create(
      <RadioButtons
        values={values}
        name='radioButtonsTest'
        onChange={onChange}
      />
    )

    expect(actual).toMatchSnapshot()
  })
})