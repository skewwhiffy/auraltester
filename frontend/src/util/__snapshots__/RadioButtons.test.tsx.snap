// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtons should match snapshot 1`] = `
<div
  onChange={[Function]}
>
  <div
    className="form-check"
  >
    <input
      className="form-check-input"
      defaultChecked={true}
      disabled={false}
      id="radioButtonsTest-0-key"
      name="radioButtonsTest"
      type="radio"
      value="0-value"
    />
    <label
      className="form-check-label"
      htmlFor="radioButtonsTest-0-key"
      title=""
    >
      0-label
    </label>
  </div>
  <div
    className="form-check"
  >
    <input
      className="form-check-input"
      defaultChecked={false}
      disabled={false}
      id="radioButtonsTest-1-key"
      name="radioButtonsTest"
      type="radio"
      value="1-value"
    />
    <label
      className="form-check-label"
      htmlFor="radioButtonsTest-1-key"
      title=""
    >
      1-label
    </label>
  </div>
  <div
    className="form-check"
  >
    <input
      className="form-check-input"
      defaultChecked={false}
      disabled={false}
      id="radioButtonsTest-2-key"
      name="radioButtonsTest"
      type="radio"
      value="2-value"
    />
    <label
      className="form-check-label"
      htmlFor="radioButtonsTest-2-key"
      title=""
    >
      2-label
    </label>
  </div>
  <div
    className="form-check"
  >
    <input
      className="form-check-input"
      defaultChecked={false}
      disabled={false}
      id="radioButtonsTest-3-key"
      name="radioButtonsTest"
      type="radio"
      value="3-value"
    />
    <label
      className="form-check-label"
      htmlFor="radioButtonsTest-3-key"
      title=""
    >
      3-label
    </label>
  </div>
  <div
    className="form-check"
  >
    <input
      className="form-check-input"
      defaultChecked={false}
      disabled={false}
      id="radioButtonsTest-4-key"
      name="radioButtonsTest"
      type="radio"
      value="4-value"
    />
    <label
      className="form-check-label"
      htmlFor="radioButtonsTest-4-key"
      title=""
    >
      4-label
    </label>
  </div>
</div>
`;
